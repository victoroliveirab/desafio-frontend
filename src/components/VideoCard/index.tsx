import Avatar from '@mui/material/Avatar';
import { useNavigate } from 'react-router-dom';
import { YoutubeVideo } from 'api/videos';
import {
  formatViewCount,
  getBestResolutionThumbUrl,
} from 'shared/helpers/videos';
import Logo from 'assets/images/logo.svg';
import styles from './styles.module.css';

interface IVideoCard {
  videoInfo: YoutubeVideo;
}

function VideoCard({ videoInfo }: IVideoCard) {
  const navigate = useNavigate();
  const thumbnailSource = getBestResolutionThumbUrl(
    videoInfo.snippet.thumbnails
  );
  const views = `${formatViewCount(videoInfo.statistics.viewCount)} views`;
  const watchVideo = () => {
    navigate(`/videos/${videoInfo.id}`, {
      state: {
        videoInfo,
      },
    });
  };
  return (
    <div className={styles['video-container']}>
      <div data-testid="thumbnail-wrapper" onClick={watchVideo}>
        <img
          src={thumbnailSource}
          className={styles['video-thumb']}
          alt={videoInfo.snippet.title}
        />
      </div>
      <div className={styles['video-info']}>
        <Avatar alt="channel" src={Logo} />
        <div className={styles['video-info__text']}>
          <h2>{videoInfo.snippet.title}</h2>
          <h4>{`${videoInfo.snippet.channelTitle} â€¢ ${views}`}</h4>
        </div>
      </div>
    </div>
  );
}

export default VideoCard;
