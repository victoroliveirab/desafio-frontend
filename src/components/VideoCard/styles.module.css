.video-container:not(:last-child) {
  width: 100%;
  height: auto;
  margin-bottom: 20px;
}

@media screen and (min-width: 576px) {
  .video-container {
    align-self: self-start;
  }
}

.video-thumb-wrapper {
  width: 100%;
  height: auto;
  margin-bottom: 10px;
}

.video-thumb {
  width: 100%;
  height: 100%;
  object-fit: cover;
}

@media screen and (min-width: 576px) {
  .video-thumb-wrapper {
    height: 250px;
    object-fit: cover;
  }
}

.video-info {
  display: flex;
  width: 100%;
  line-height: 2;
}

.video-info > * + * {
  margin-left: 8px;
}

.video-info__text {
  height: 100%;
  min-height: 40px;
}

.video-info h2 {
  font-size: 1rem;
  display: -webkit-box;
  -webkit-box-orient: vertical;
  text-overflow: ellipsis;
  -webkit-line-clamp: 2;
  overflow: hidden;
  margin: 0;
  max-height: 40px;
  line-height: 1.2;
  font-weight: normal;
}

.video-info h4 {
  font-size: 0.875rem;
  margin: 0;
  font-weight: normal;
}
