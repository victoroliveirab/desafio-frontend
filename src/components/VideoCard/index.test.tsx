import { fireEvent, render, RenderResult } from 'test-utils';
import videoInfo from '__mocks__/youtube-video';
import VideoCard from '.';

const mockUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as any),
  useNavigate: () => mockUseNavigate,
}));

let component: RenderResult;

describe('<VideoCard />', () => {
  beforeEach(() => {
    component = render(<VideoCard videoInfo={videoInfo} />);
  });
  test('should match snapshot', () => {
    expect(component).toMatchSnapshot();
  });
  test('should navigate when thumbnail is clicked', () => {
    const thumbnailWrapper = component.getByTestId('thumbnail-wrapper');

    fireEvent.click(thumbnailWrapper);

    expect(mockUseNavigate).toBeCalled();
  });
});
