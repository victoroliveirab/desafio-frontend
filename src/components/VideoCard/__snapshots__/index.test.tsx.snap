// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VideoCard /> should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="video-container"
      >
        <div
          data-testid="thumbnail-wrapper"
        >
          <img
            alt="Would You Swim With Sharks For $100,000?"
            class="video-thumb"
            src="https://i.ytimg.com/vi/LnlKwzc_TNA/maxresdefault.jpg"
          />
        </div>
        <div
          class="video-info"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
          >
            <img
              alt="channel"
              class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
              src="logo.svg"
            />
          </div>
          <div
            class="video-info__text"
          >
            <h2>
              Would You Swim With Sharks For $100,000?
            </h2>
            <h4>
              MrBeast • 13.5M views
            </h4>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="video-container"
    >
      <div
        data-testid="thumbnail-wrapper"
      >
        <img
          alt="Would You Swim With Sharks For $100,000?"
          class="video-thumb"
          src="https://i.ytimg.com/vi/LnlKwzc_TNA/maxresdefault.jpg"
        />
      </div>
      <div
        class="video-info"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
        >
          <img
            alt="channel"
            class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="logo.svg"
          />
        </div>
        <div
          class="video-info__text"
        >
          <h2>
            Would You Swim With Sharks For $100,000?
          </h2>
          <h4>
            MrBeast • 13.5M views
          </h4>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
